File: src/podcast/services/tts.py
"""TTS service using Modal's GPU-accelerated Chatterbox."""

import json
import logging
import os
import uuid

import modal

from podcast.config import settings
from podcast.database import get_session
from podcast.models import Episode, PodcastSettings

logger = logging.getLogger(__name__)


def get_tts_progress(episode_id: uuid.UUID) -> dict | None:
    """Read TTS progress from the progress file, if it exists.

    Note: Modal TTS doesn't write progress files (GPU work is remote).
    Returns None since there's no segment-by-segment progress to report.
    Callers should handle gracefully (show spinner without progress).
    """
    return None


def _validate_voice_ref_path(path: str) -> str | None:
    """Validate that a voice ref path resolves to a file inside the allowed directory.

    Returns the resolved path if valid, None otherwise.
    """
    allowed_dir = os.path.realpath(settings.voice_refs_dir)
    resolved = os.path.realpath(os.path.join(allowed_dir, os.path.basename(path)))
    if not resolved.startswith(allowed_dir + os.sep) and resolved != allowed_dir:
        logger.warning("Voice ref path escapes allowed directory: %s", path)
        return None
    return resolved


def _read_voice_ref_bytes(db_path: str | None, default_filename: str) -> bytes | None:
    """Read voice reference WAV file as bytes.

    Args:
        db_path: Path from database (may be None or nonexistent)
        default_filename: Default filename in voice_refs dir (e.g. "host_a.wav")

    Returns:
        WAV bytes if file exists, None otherwise
    """
    # Try database path first (validate it stays within voice_refs_dir)
    if db_path:
        safe_path = _validate_voice_ref_path(db_path)
        if safe_path and os.path.exists(safe_path):
            logger.debug("Reading voice ref from DB path: %s", safe_path)
            with open(safe_path, "rb") as f:
                return f.read()

    # Try default path
    default_path = os.path.join(settings.voice_refs_dir, default_filename)
    if os.path.exists(default_path):
        logger.debug("Reading voice ref from default path: %s", default_path)
        with open(default_path, "rb") as f:
            return f.read()

    logger.warning(
        "No voice ref found, tried: db=%s, default=%s", db_path, default_path
    )
    return None


async def synthesize_speech(episode_id: uuid.UUID) -> dict:
    """Convert transcript segments to speech using Modal GPU TTS.

    Returns metrics dict with generation stats.
    """
    from sqlalchemy import select as sa_select

    # Read data from DB
    async with get_session() as db:
        episode = await db.get(Episode, episode_id)
        if not episode or not episode.transcript:
            raise ValueError(f"Episode {episode_id} not found or has no transcript")

        result = await db.execute(
            sa_select(PodcastSettings).where(PodcastSettings.user_id == episode.user_id)
        )
        podcast_settings = result.scalar_one_or_none()
        host_a = podcast_settings.host_a_name if podcast_settings else "Alex"
        voice_ref_a = podcast_settings.voice_ref_a_path if podcast_settings else None
        voice_ref_b = podcast_settings.voice_ref_b_path if podcast_settings else None
        user_id = episode.user_id
        segments = json.loads(episode.transcript)

    logger.info("Synthesizing %d segments for episode %s", len(segments), episode_id)

    # Read voice ref bytes
    voice_ref_a_bytes = _read_voice_ref_bytes(voice_ref_a, "host_a.wav")
    voice_ref_b_bytes = _read_voice_ref_bytes(voice_ref_b, "host_b.wav")

    # Call Modal function
    try:
        fn = modal.Function.from_name("podcast-tts", "generate_tts")
        result = await fn.remote.aio(
            segments=segments,
            host_a_name=host_a,
            voice_ref_a_bytes=voice_ref_a_bytes,
            voice_ref_b_bytes=voice_ref_b_bytes,
        )
    except Exception as e:
        logger.error("Modal TTS call failed: %s", e)
        raise

    # Write WAV bytes to disk (namespaced per user)
    user_audio_dir = os.path.join(settings.audio_dir, str(user_id))
    output_wav = os.path.join(user_audio_dir, f"{episode_id}.wav")
    os.makedirs(user_audio_dir, exist_ok=True)
    with open(output_wav, "wb") as f:
        f.write(result["wav_bytes"])
    logger.info("Wrote audio file: %s", output_wav)

    return result["metrics"]

File: src/services/tts.ts
import fs from "node:fs";
import path from "node:path";
import { eq } from "drizzle-orm";
import { db } from "../database.js";
import { episodes, podcastSettings } from "../schema.js";
import { settings } from "../config.js";
import { createLogger } from "../log-handler.js";

const log = createLogger("tts");

export function getTtsProgress(_episodeId: string): null {
  // Modal TTS doesn't write progress files (GPU work is remote).
  return null;
}

function validateVoiceRefPath(refPath: string): string | null {
  const allowedDir = fs.realpathSync(settings.voiceRefsDir);
  const resolved = fs.realpathSync(
    path.join(allowedDir, path.basename(refPath))
  );
  if (!resolved.startsWith(allowedDir + path.sep) && resolved !== allowedDir) {
    log.warn("Voice ref path escapes allowed directory: %s", refPath);
    return null;
  }
  return resolved;
}

function readVoiceRefBytes(
  dbPath: string | null | undefined,
  defaultFilename: string
): Buffer | null {
  // Try database path first
  if (dbPath) {
    try {
      const safePath = validateVoiceRefPath(dbPath);
      if (safePath && fs.existsSync(safePath)) {
        return fs.readFileSync(safePath);
      }
    } catch {}
  }
  // Try default path
  const defaultPath = path.join(settings.voiceRefsDir, defaultFilename);
  if (fs.existsSync(defaultPath)) {
    return fs.readFileSync(defaultPath);
  }
  log.warn("No voice ref found, tried: db=%s, default=%s", dbPath, defaultPath);
  return null;
}

export async function synthesizeSpeech(
  episodeId: string
): Promise<Record<string, unknown>> {
  // Read data from DB
  const [episode] = await db
    .select()
    .from(episodes)
    .where(eq(episodes.id, episodeId))
    .limit(1);
  if (!episode || !episode.transcript) {
    throw new Error(
      `Episode ${episodeId} not found or has no transcript`
    );
  }

  const [psettings] = await db
    .select()
    .from(podcastSettings)
    .where(eq(podcastSettings.userId, episode.userId))
    .limit(1);

  const hostA = psettings?.hostAName ?? "Alex";
  const voiceRefA = psettings?.voiceRefAPath ?? null;
  const voiceRefB = psettings?.voiceRefBPath ?? null;
  const segments = JSON.parse(episode.transcript);

  log.info(
    "Synthesizing %d segments for episode %s",
    segments.length,
    episodeId
  );

  // Read voice ref bytes
  const voiceRefABytes = readVoiceRefBytes(voiceRefA, "host_a.wav");
  const voiceRefBBytes = readVoiceRefBytes(voiceRefB, "host_b.wav");

  // Call Modal TTS via HTTP web endpoint
  const modalUrl = settings.modalTtsUrl;
  if (!modalUrl) {
    throw new Error(
      "MODAL_TTS_URL is not configured. Deploy modal_app.py and set the URL."
    );
  }

  const body: Record<string, unknown> = {
    segments,
    host_a_name: hostA,
    voice_ref_a_bytes: voiceRefABytes
      ? voiceRefABytes.toString("base64")
      : null,
    voice_ref_b_bytes: voiceRefBBytes
      ? voiceRefBBytes.toString("base64")
      : null,
  };

  const res = await fetch(modalUrl, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(1800_000), // 30 min timeout
  });

  if (!res.ok) {
    const errText = await res.text();
    throw new Error(`Modal TTS call failed (${res.status}): ${errText}`);
  }

  const result = (await res.json()) as {
    wav_bytes: string;
    metrics: Record<string, unknown>;
  };

  // Write WAV bytes to disk (namespaced per user)
  const userAudioDir = path.join(settings.audioDir, episode.userId);
  const outputWav = path.join(userAudioDir, `${episodeId}.wav`);
  fs.mkdirSync(userAudioDir, { recursive: true });
  fs.writeFileSync(outputWav, Buffer.from(result.wav_bytes, "base64"));
  log.info("Wrote audio file: %s", outputWav);

  return result.metrics;
}
